wagtail/wagtailcore/permission_policies/collections.py

```diff
@@ -0,0 +1,348 @@
+        that apply to that collection.
+        """
+        return get_user_model().objects.filter(
+            self._users_with_perm_filter(actions, collection=collection)
+        ).distinct()
+
+    def collections_user_has_permission_for(self, user, action):
+        """
+        Return a queryset of all collections in which the given user has
+        permission to perform the given action
+        """
+        return self.collections_user_has_any_permission_for(user, [action])
+
+
+class CollectionPermissionPolicy(CollectionPermissionLookupMixin, BaseDjangoAuthPermissionPolicy):
+    """
+    A permission policy for objects that are assigned locations in the Collection tree.
+    Permissions may be defined at any node of the hierarchy, through the
+    GroupCollectionPermission model, and propagate downwards. These permissions are
+    applied to objects according to the standard django.contrib.auth permission model.
+    """
+    def user_has_permission(self, user, action):
+        """
+        Return whether the given user has permission to perform the given action
+        on some or all instances of this model
+        """
+        return self._check_perm(user, [action])
+
+    def user_has_any_permission(self, user, actions):
+        """
+        Return whether the given user has permission to perform any of the given actions
+        on some or all instances of this model
+        """
+        return self._check_perm(user, actions)
+
+    def users_with_any_permission(self, actions):
+        """
+        Return a queryset of users who have permission to perform any of the given actions
+        on some or all instances of this model
+        """
+        return self._users_with_perm(actions)
+
+    def user_has_permission_for_instance(self, user, action, instance):
+        """
+        Return whether the given user has permission to perform the given action on the
+        given model instance
+        """
+        return self._check_perm(user, [action], collection=instance.collection)
+
+    def user_has_any_permission_for_instance(self, user, actions, instance):
+        """
+        Return whether the given user has permission to perform any of the given actions
+        on the given model instance
+        """
+        return self._check_perm(user, actions, collection=instance.collection)
+
+    def instances_user_has_any_permission_for(self, user, actions):
+        """
+        Return a queryset of all instances of this model for which the given user has
+        permission to perform any of the given actions
+        """
+        if not (user.is_active and user.is_authenticated()):
+            return self.model.objects.none()
+        elif user.is_superuser:
+            return self.model.objects.all()
+        else:
+            # filter to just the collections with this permission
+            return self.model.objects.filter(
+                collection__in=list(self._collections_with_perm(user, actions))
```

Comment:
I would definitely have preferred to do it with a single query, but unfortunately this was necessary for the search mechanism to work. We call `search(query_string)` on the returned queryset - this involves converting the query filter criteria into a search query (to be passed to either elasticsearch or our ORM-based search backend), and one of the rules for this is that we can only filter on fields that have been indexed for search (using `index.FilterField`). This means we can't perform arbitrary complex subqueries here - but we can do `__in` lookups on the 'collection' field, as that field has been indexed.